### ISSUE TITLE
Training vs tensorboard metrics

### ISSUE THREAD

**Issue by smlkdev** [myshell-ai/MeloTTS#211]
Will my training yield better results over time? Currently, the training took about 9 hours.
I have 1500 wav samples, with a total audio length of approximately 2 hours.

![Screenshot 2024-11-08 at 11 53 27](https://github.com/user-attachments/assets/82c7cc5f-00ab-4bf8-9220-a17d651b42fe)

What other metrics should I pay attention to in TensorBoard?

**Comment by smlkdev** [comment]
Update after ~34h:
Little improvement visible but note sure should I keep it longer because of the flattening. 

![Screenshot 2024-11-09 at 10 41 25](https://github.com/user-attachments/assets/2c7ebdcb-5244-400a-b14e-449c89b400d1)
![Screenshot 2024-11-09 at 10 41 41](https://github.com/user-attachments/assets/c5cb6ebd-b929-41c1-8b2b-e0a0476b8087)

**Comment by jeremy110** [comment]
We usually look at g/total, and from your graph, it seems to be decreasing pretty well. But I’m not sure if 2 hours of training data is enough; I initially used around 8 to 10 hours for training.

**Comment by smlkdev** [comment]
> We usually look at g/total, and from your graph, it seems to be decreasing pretty well. But I’m not sure if 2 hours of training data is enough; I initially used around 8 to 10 hours for training.

@jeremy110 Thank you for your response! I’m honestly a bit hooked on watching the progress as it keeps going down, so I can’t seem to stop checking in :-)

Currently at 68 hours.

![Screenshot 2024-11-10 at 22 05 29](https://github.com/user-attachments/assets/708b7b1c-9a52-4ae2-880b-1dea8be0159c)

I’m planning to create an 8-10 hour audio dataset for the next training session. Could you suggest what kind of text data I should gather for it? So far, I’ve used random articles and some ChatGPT-generated data, but I’ve heard that people sometimes read books, for example. Is there perhaps a dataset available with quality English sentences that covers a variety of language phenomena? I tried to find it but no results.

**Comment by jeremy110** [comment]
@smlkdev 
Basically, this training can be kept short since it’s just a fine-tuning session; no need to make it too long. Here’s my previous tensorboard log for your reference(https://github.com/myshell-ai/MeloTTS/issues/120#issuecomment-2105728981).

I haven’t specifically researched text types. My own dataset was professionally recorded, with sentences that resemble reading books. I’m not very familiar with English datasets—are you planning to train in English?

**Comment by smlkdev** [comment]
This is my first attempt with ML/training/voice cloning and decided to use english. I read briefly Thai thread and it was way too complex for me to start with.

Your training was 32 hours long and for me (I'm not the expert) infer voice matched original :) That's really nice. Is it a voice that had 8-10 hours of audio as you mentioned earlier?

**Comment by jeremy110** [comment]
Yes, that's correct. I tried both single-speaker and multi-speaker models, and the total duration is around 8-10 hours.

If this is your first time getting into it, I recommend you try F5-TTS. There are a lot of people in the forums who have trained their own models, and some even wrote a Gradio interface, which is very convenient.

**Comment by smlkdev** [comment]
@jeremy110 thank your for your responses.

Is F5-TTS better than MeloTTS in terms of quality?

I just realized that my cloned MeloTTS voice doesn’t add breaks between sentences. I have to add them manually—by splitting the text into sentences, breaking it down into smaller parts, generating and then merging it back together after adding pauses. It can be made automatically of course but still a bit of work.

**Comment by jeremy110** [comment]
In terms of quality, I think F5-TTS is quite good. You can try it out on the Hugging Face demo.

The pauses within sentences mainly depend on your commas (","). The program adds a space after punctuation to create a pause. However, if the audio files you trained on have very little silence before and after the speech, the generated audio will also have little silence. Of course, you can add the pauses manually, but you could also address it by adjusting the training data.

**Comment by kadirnar** [comment]
@smlkdev I am training the melotts model with sentiment data. But I couldn't get the tensorboard graphs to work. Can you share your sample code?

**Comment by kadirnar** [comment]
These are written in the train.log file. Training is still ongoing. Are these important?

```bash
2024-11-19 09:22:10,339	example	ERROR	enc_p.language_emb.weight is not in the checkpoint
2024-11-19 09:22:10,340	example	ERROR	emb_g.weight is not in the checkpoint
```

**Comment by manhcuong17072002** [comment]
@jeremy110 Hello, I would like to inquire about the data preparation process when training on multiple speakers. Is it necessary for each speaker to have a comparable amount of data? For instance, if Speaker A has 10 hours of audio and Speaker B only has 1 hour, is it possible to create a good model, or does Speaker B also require approximately 10 hours of audio? Thank you

**Comment by jeremy110** [comment]
@manhcuong17072002 Hello~
In my training, some speakers had 1 or 2 hours of audio, while others had 30 minutes, and in the end, there were about 10 hours of total data. I was able to train a decent model, but for speakers with less data, their pronunciation wasn't as accurate.

**Comment by manhcuong17072002** [comment]
@jeremy110 Oh, if that's the case, that's wonderful. Collecting data and training the model will become much easier with your idea. So, when training, you must have used many speaker IDs, right? And do you find their quality sufficient for deployment in a real-world environment? I'm really glad to hear your helpful feedback. Thank you very much!

**Comment by jeremy110** [comment]
@manhcuong17072002 

Yes, there are about 15 speakers. Of course, if you have enough people, you can continue to increase the number. After 10 hours, the voice quality is quite close, but if you want better prosody, you might need more speakers and hours.

From the TTS systems I've heard, the voice quality is about above average, but when it comes to deployment, you need to consider inference time. For this, MeloTTS is quite fast.

**Comment by manhcuong17072002** [comment]
@jeremy110 Thank you for the incredibly helpful information. Let me summarize a few points:
- Training data: 8 to 10 hours of audio is sufficient to train the MeloTTS model, and more data is always welcome.
- Number of speakers: 15. 30 minutes to 2 hours of data per speaker yields good results. More data per speaker leads to better results.
- Deployment speed: MeloTTS is relatively fast to deploy.

However, I've experimented with various TTS models and noticed that if the text isn't broken down into smaller chunks, the generated speech quality degrades towards the end of longer passages. Have you tested this with MeloTTS? If so, could you share your experimental process? Thank you so much.

**Comment by jeremy110** [comment]
@manhcuong17072002 
You're welcome, your conclusion is correct.


Normally, during training, long audio files are avoided to prevent GPU OOM (Out of Memory) issues. Therefore, during inference, punctuation marks are typically used to segment the text, ensuring that each sentence is closer to the length used during training for better performance. MeloTTS performs this segmentation based on punctuation during inference, and then concatenates the individual audio files after synthesis.

**Comment by manhcuong17072002** [comment]
@jeremy110  I'm so sorry but I suddenly have a question about training on a multiple speakers dataset. Is it possible for Speaker A to pronounce words that exist in other speakers but not in A? Because if not, dividing the dataset into multiple speakers would be pointless and the model would not be able to cover the entire vocabulary of a language. Have you tried this before and what are your thoughts on this? Thank you.

**Comment by jeremy110** [comment]
@manhcuong17072002 
If we consider 30 minutes of audio, assuming each word takes about 0.3 seconds, there would be around 5000–6000 words. These words would then be converted into phoneme format, meaning they would be broken down into their phonetic components for training. With 6000 words, the model would learn most of the phonemes. However, when a new word is encountered, it will be broken down into the phonemes it has already learned. I haven't done rigorous testing, but in my case, the model is able to produce similar sounds.

**Comment by manhcuong17072002** [comment]
@jeremy110  Thanks for your useful information

**Comment by smlkdev** [comment]
@jeremy110 Even if latest messages were not addressed directly to me - I want to thank you as well, you are giving me a fresh point of view how to look at dataset.

1. Using your suggestion, I'm currently testing how F5-TTS works and what effects it produces. Here are my charts (same dataset as in this thread). Should I stop fine-tuning now that the LR has dropped so low or can it still improve nicely?

![Screenshot 2024-11-22 at 18 36 31](https://github.com/user-attachments/assets/fe845461-2d68-4036-979b-397f5d6b42fe)

2. And how should I choose the right/best checkpoint? Is it a matter of listening to the dialogue, or should I rely on the loss from the chart, i.e., lower = better? Of course it is hard to get very exact checkpoint when I'm doing save every 2.5k steps.

3. If I wanted to create a short dataset from scratch, how many times should each word appear in the audio recordings to make it meaningful? Would just once be enough? I imagine that if I were creating a dataset to perform exceptionally well in a specific field, like "cooking," I could build one using words more frequently used in that domain, such as "flour," "knife," or "tomato." I’m also guessing that the sentences in the dataset should include the most commonly used words in English along with niche-specific words, so my model performs as well as possible in that area (though it might struggle in something like "aviation" niche).

**Comment by jeremy110** [comment]
@smlkdev 
1. Basically, fine-tuning usually requires less than 10 epochs; the settings from others are used as a reference.
![image](https://github.com/user-attachments/assets/46c400cb-41ca-4930-8a28-f54aa8223d1b)
2. Typically, a model trained for 10 epochs is already good for use, and the last one is usually the best.
3. If you're training a new language, a small dataset is not ideal; it still needs to be of a certain length to achieve better learning.
Since I'm not sure if you're training a new language, if you are, I tested small datasets and found that around 20–40 hours of audio is necessary, with each clip lasting 2–10 seconds. It can be either multiple speakers or a single speaker, but zero-shot performance is poor with a single speaker. Additionally, I tested with 350 hours of audio and about 300 speakers, and the zero-shot performance was excellent. Using my own voice as a reference, the synthesized voice was very close to my own. Finally, for F5-TTS to generate good audio, the reference is crucial. In my tests, 6–10 second clips worked best. So, for your third question, you can include these specific words in your reference, and the inference quality for this domain will improve.

**Comment by kadirnar** [comment]
@jeremy110 

https://huggingface.co/datasets/reach-vb/jenny_tts_dataset I am training this dataset. I made 10 epochs from the config settings. There is epoch 30 in the train.log file and it is still training. What should I fix?

How do you think the graphs are? Can you interpret them?

<img width="1145" alt="image" src="https://github.com/user-attachments/assets/09280a45-9bd9-4842-b77c-47fc151ac140">

**Comment by jeremy110** [comment]
@kadirnar hello~
Typically, the parameters would be read from the YAML file, but it's okay. You can stop it at the appropriate time. I’ve kind of forgotten how many steps I originally set for training, but you can refer to my TensorBoard.

I’m not very experienced in the TTS field, but for MeloTTS, it mainly uses loss functions common in GANs, involving the Discriminator and Generator. Personally, I check the loss/g/total and loss/g/mel to assess whether the training results are as expected.

From your graphs, since there is no loss/g/total, I cannot judge the result. From my own training, the values typically range from 45 to 60, depending on your dataset.

**Comment by manhcuong17072002** [comment]
@jeremy110 What do you think if I have a small dataset and I combine the audios in the dataset to create a larger one? For example, initially we have 2 audios A and B. We combine the 2 audios as follows: Audio A + Audio B = Audio C, from which we get 3 audios A, B, C. Do you think this will significantly affect the training results compared to the original small dataset? Thank you.

**Comment by jeremy110** [comment]
@manhcuong17072002 

This approach can indeed enhance the data and may provide a slight improvement, but several points need to be considered. Since MeloTTS uses BERT to extract feature vectors, if we randomly concatenate the text of two audio files and then extract feature vectors, can it still effectively represent the prosody of the text?
You could try it out and see how it performs.

**Comment by kadirnar** [comment]
@jeremy110 Thanks for your help. The sound quality is better now.  I am training with a larger dataset. The train.log file says these error messages. How can I fix this?

```bash
2024-11-26 18:12:12,084	example	ERROR	enc_p.encoder.attn_layers.0.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,084	example	ERROR	enc_p.encoder.attn_layers.0.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,084	example	ERROR	enc_p.encoder.attn_layers.1.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,084	example	ERROR	enc_p.encoder.attn_layers.1.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,084	example	ERROR	enc_p.encoder.attn_layers.2.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,084	example	ERROR	enc_p.encoder.attn_layers.2.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,084	example	ERROR	enc_p.encoder.attn_layers.3.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,084	example	ERROR	enc_p.encoder.attn_layers.3.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,084	example	ERROR	enc_p.encoder.attn_layers.4.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,084	example	ERROR	enc_p.encoder.attn_layers.4.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,084	example	ERROR	enc_p.encoder.attn_layers.5.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,084	example	ERROR	enc_p.encoder.attn_layers.5.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.0.enc.attn_layers.0.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.0.enc.attn_layers.0.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.0.enc.attn_layers.1.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.0.enc.attn_layers.1.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.0.enc.attn_layers.2.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.0.enc.attn_layers.2.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.2.enc.attn_layers.0.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.2.enc.attn_layers.0.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.2.enc.attn_layers.1.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.2.enc.attn_layers.1.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.2.enc.attn_layers.2.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.2.enc.attn_layers.2.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.4.enc.attn_layers.0.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.4.enc.attn_layers.0.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.4.enc.attn_layers.1.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.4.enc.attn_layers.1.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.4.enc.attn_layers.2.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.4.enc.attn_layers.2.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.6.enc.attn_layers.0.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.6.enc.attn_layers.0.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.6.enc.attn_layers.1.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.6.enc.attn_layers.1.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.6.enc.attn_layers.2.emb_rel_k is not in the checkpoint
2024-11-26 18:12:12,085	example	ERROR	flow.flows.6.enc.attn_layers.2.emb_rel_v is not in the checkpoint
2024-11-26 18:12:12,117	example	INFO	Loaded checkpoint './logs/example/G_0.pth' (iteration 5)
2024-11-26 18:12:12,230	example	INFO	Loaded checkpoint './logs/example/D_0.pth' (iteration 5)
2024-11-26 18:24:00,697	example	INFO	Train Epoch: 5 [94%]
2024-11-26 18:24:00,698	example	INFO	[2.5422043800354004, 2.1978163719177246, 8.537149429321289, 20.567224502563477, 2.303469181060791, 2.3287158012390137, 400, 9.99875e-05]
2024-11-26 18:24:21,757	example	INFO	====> Epoch: 5
2024-11-26 18:31:56,010	example	INFO	====> Epoch: 6
```

**Comment by yukiarimo** [comment]
How to get the tensorboard metrics when I started the training?
